### ISSUE TITLE
Support schema when calling `yamldecode`

### ISSUE THREAD

**Issue by minherz** [hashicorp/terraform#31969]
### Terraform Version

```shell
Terraform v1.3.1
on linux_amd64
```

### Use Cases

Provisioning a configuration based deployment often requires using YAML files which are provided by a user. In these cases a path to the files is provided as an input variable. While it should be sometimes possible to validate that the provided YAML files are properly formatted and match the expected schema, it is not always feasible. The `yamldecode` function validates YAML format.
It would be great if it will accept as a second, optional argument a path to another YAML file or a string that describes the schema vs which the read file should be validate.

### Attempted Solutions

There are no built-in that support this functionality. I did not find third party modules that implement it either.

### Proposal

Because Terraform HCL does not support argument extensions and function overloads, a new function `yamlvalidate` can be introduced. The function accepts two arguments:

- A string that is expected to be a subset of YAML or a path to the YAML file
- A string containing a YAML schema for validation or a path to the file that containing the schema

The function can return the YAML that was passed in the first argument in the case of successful validation to allow it to be used with `yamldecode`.

**Comment by apparentlymart** [comment]
Hi @minherz! Thanks for sharing this feature-request.

Is there an existing YAML schema format that would be appropriate to use for this? I don't think we'd like to invent a Terraform-specific schema language just for this function, but if there is an existing schema language in common use (which ideally has an existing implementation written in Go) then it could be reasonable to support this.

With that said, we also intend to focus only on very common use-cases for internal functions moving forward, and intend to delegate more complex needs to provider plugins. Currently that means writing a provider which contains a data source which only does internal computation, although we are considering allowing provider plugins to contribute normal functions to the language in future too, which would allow implementing exactly what you've described as a third-party extension. My initial instinct is that this request would be better met by a third-party provider rather than by a core language feature, because it seems like a significant amount of additional functionality which is not a core part of Terraform's mission.

**Comment by minherz** [comment]
Hi @apparentlymart , there are official [recommendations about schemas][1]. AFAIK many languages use implementations based on JSON-like schema format. I understand your arguments about delegating non-core functions to providers. It makes a lot of sense and I saw this working very well with kubernetes project. A minor contra-argument here is that yaml and json decode/encode functionality is already there. Meaning it already performs a certain level of the content validation. However, I agree that the quickest way to do it would be implementing a provider or even a custom module.

[1]: https://yaml.org/spec/1.2.2/#chapter-10-recommended-schemas

**Comment by apparentlymart** [comment]
Indeed... YAML encoding and decoding is a sufficiently-common case that it was pragmatic to include it in the core language. It's not yet clear that YAML _schema validation_ is similarly common, since this has been the first request for it since we added `yamldecode` in Terraform v0.12.

That doesn't mean we cannot possibly add it if it proves to be a common problem with a common shared solution, but I just wanted to be explicit about how we tend to evaluate requests like this.

The fact that there are several different approaches also tends to suggest it's better to delegate to plugins, because that way many different approaches can coexist in different providers and authors can choose whichever works best for their situation. Otherwise we'd inevitably need to somehow decide which of the many schema designs is the "best", because we don't have the resources to support many different YAML schema implementations. (Whether we have the resources to support even _one_ still remains to be seen, of course!)

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!
